File: 2022/deno-aoc-2022/src/day-04.ts
const day04a = async () => {
  const inputFile = await Deno.readTextFile("input/day-04.txt");
  const lines = inputFile.split("\n", -1);

  let count = 0;

  for (const line of lines) {
    const segments = line.split(",", -1);
    const span0 = segments[0].split("-", -1).map((s) => Number(s));
    const span1 = segments[1].split("-", -1).map((s) => Number(s));

    if (
      (span0[0] >= span1[0] && span0[1] <= span1[1]) ||
      (span1[0] >= span0[0] && span1[1] <= span0[1])
    ) {
      count++;
    }
  }

  console.log(`Day 4a total ${count}`);
};

const day04b = async () => {
  const inputFile = await Deno.readTextFile("input/day-04.txt");
  const lines = inputFile.split("\n", -1);

  let count = 0;

  for (const line of lines) {
    const segments = line.split(",", -1);
    const span0 = segments[0].split("-", -1).map((s) => Number(s));
    const span1 = segments[1].split("-", -1).map((s) => Number(s));

    if (span0[1] >= span1[0] && span0[0] <= span1[1]) {
      count++;
    }
  }

  console.log(`Day 4b total ${count}`);
};

if (import.meta.main) {
  await day04a();
  await day04b();
}

File: 2023/swift-aoc-2023/Sources/day04.swift
import Foundation
import Parsing

func parseCard(cardString: String) -> (Int, [Int], [Int])? {
    let parser = Parse {
        "Card"
        Whitespace()
        Int.parser()
        ":"
        Many {
            Whitespace()
            Int.parser()
        }
        " |"
        Many {
            Whitespace()
            Int.parser()
        }
    }

    return try? parser.parse(cardString)
}

func day04a() {
    if let contents = try? String(contentsOfFile: "input/day-04.txt") {
        let lines = contents.components(separatedBy: "\n")

        let result = lines.reduce(0) {
            if let card = parseCard(cardString: $1) {
                let winningNumbers = Set(card.1)
                return $0 + card.2.reduce(0) {
                    winningNumbers.contains($1) ? ($0 > 0 ? $0 * 2 : 1) : $0
                }
            } else {
                return 0
            }
        }
        print(result)
    }
}

func day04b() {
    if let contents = try? String(contentsOfFile: "input/day-04.txt") {
        let lines = contents.components(separatedBy: "\n")

        var scores: [Int: Int] = [:]

        for line in lines {
            if let card = parseCard(cardString: line) {
                let winningNumbers = Set(card.1)
                scores[card.0] = card.2.reduce(0) {
                    winningNumbers.contains($1) ? $0 + 1 : $0
                }
            }
        }

        var cards = Array(1 ... 214)
        var process = 0

        while process < cards.count {
            let card = cards[process]
            if let wins = scores[card] {
                if wins > 0 {
                    cards.append(contentsOf: Array((card + 1) ... (card + wins)))
                }
            }
            process += 1
        }

        print(cards.count)
    }
}
